src/hotspot/os/linux/os_linux.cpp

```diff
@@ -816,6 +819,32 @@ static size_t get_static_tls_area_size(const pthread_attr_t *attr) {
   return tls_size;
 }
 
+// In glibc versions prior to 2.27 the guard size mechanism
+// was not implemented properly. The posix standard requires adding
+// the size of the guard pages to the stack size, instead glibc
+// took the space out of 'stacksize'. Thus we need to adapt the requested
+// stack_size by the size of the guard pages to mimic proper behaviour.
+// The fix in glibc 2.27 has now been backported to numerous earlier
+// glibc versions so we need to do a dynamic runtime check.
+bool os::Linux::AdjustStackSizeForGuardPages = true;
+
+static void init_adjust_stack_for_guard_pages() {
+  get_minstack_init();
+  if (_get_minstack_func != nullptr) {
+    pthread_attr_t attr;
+    pthread_attr_init(&attr);
+    size_t min_stack = _get_minstack_func(&attr);
+    pthread_attr_setguardsize(&attr, 16*K );
```

Comment:
Does this work on arches with 64K pages? Perhaps we need to check the return value of this call to make sure we were successful in setting the guard size to begin with. Otherwise we can "accidentally" discover the min_stack is not changing?

Reply:
Good point. Should I use `os::vm_page_size()`?

The call cannot actually fail on Linux (none of these call can, hence no return code checking).

Reply:
@tstuefe  that is all true and correct, but what we don't know (without looking at the implementation) is what `__get_minstack` will do with the value in the attribute. I guess I need to find the old version of the code and see whether it does a raw read or adjusts it somehow based on pagesize.

Reply:
The value is used as-is so its actual value does not matter.